### PR TITLE
Always set a crumb issuer

### PR DESCRIPTION
**Pull request by daniel-beck**
Twitter driven development: https://twitter.com/danielbeck/status/1214346538618621964

Make sure that the crumb issuer is always set.

#### The default

> Screenshot

#### A custom crumb issuer

> Screenshot

#### With escape hatch set

> Screenshot

#### A note on not removing the UI

Configuration UI for some options in Jenkins are not shown when there's only one option. I decided against this approach here for two reasons:

* Even with the default crumb issuer, there's still the proxy compatibility setting.
* The configuration disappearing altogether might worry experienced and security-conscious Jenkins administrators.

### On proxy compatibility

I decided to use the same approach to proxy compatibility as is currently used in the setup wizard, i.e. require a system property to enable greater compatibility out of the box.

In some cases (CSRF protection disabled, load balancer or otherwise unusually configured reverse proxy in front of Jenkins), this may result in a problem requiring a minor system property dance, bypassing the proxy for a moment to reconfigure, or patching `config.xml` to restore UI functionality in Jenkins.

I think these cases are rare enough so this shouldn't be a huge problem; an argument can be made to slightly lower protection for greater compatibility: Since https://jenkins.io/security/advisory/2019-07-17/#SECURITY-626, we make the the session ID part of the crumb, so requiring an IP address match on top of that isn't _that_ important.

### Escape hatch

While there's an escape hatch in `hudson.security.csrf.GlobalCrumbIssuerConfiguration.DISABLE_CSRF_PROTECTION`, I wouldn't advertise it much.

How it works:

* If the system property is set on Jenkins startup, Jenkins will set or unset the crumb issuer according to the value set.
* While Jenkins is running, the config UI can be locked/unlocked by setting the field of the same name as the system property to `true`/`false`. This will not immediately change the configured crumb issuer; a global security config form submission is needed for that.

### Proposed changelog entries

* Remove the ability to have CSRF protection disabled. Instances upgrading from older versions of Jenkins will have CSRF protection enabled and the default issuer set if they currently have it disabled.

### Proposed upgrade guidelines

Jenkins will automatically enable CSRF protection with the default crumb issuer if it was disabled before. The ability to not have CSRF protection enabled has been deprecated and removed from the UI. Set the system property `hudson.security.csrf.GlobalCrumbIssuerConfiguration.DISABLE_CSRF_PROTECTION` to `true` on startup to disable CSRF protection as well as the configuration UI for it.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

**Review by Wadeck — APPROVED** [review]
Not manually tested, nothing seems wrong. Explanations are good enough, 👍

**Comment by jglick** [comment]
> Configuration UI for some options in Jenkins are not shown when there's only one option.

Which is bad UI I think. Better to be consistent and show the one option.

File: core/src/main/java/hudson/security/csrf/GlobalCrumbIssuerConfiguration.java [hunk 1]
```diff
@@ -49,14 +52,31 @@
...
         Jenkins j = Jenkins.get();
-        if (json.has("csrf")) {
-            JSONObject csrf = json.getJSONObject("csrf");
-            j.setCrumbIssuer(CrumbIssuer.all().newInstanceFromRadioList(csrf, "issuer"));
+
+        if (json.has("crumbIssuer")) {
+            j.setCrumbIssuer(req.bindJSON(CrumbIssuer.class, json.getJSONObject("crumbIssuer")));
         } else {
-            j.setCrumbIssuer(null);
+            // set the default crumb issuer
```
Comment (jglick) [inline core/src/main/java/hudson/security/csrf/GlobalCrumbIssuerConfiguration.java]:
```suggestion
```

Adds nothing to what you would understand from reading the code.

File: core/src/main/java/hudson/security/csrf/GlobalCrumbIssuerConfiguration.java [hunk 2]
```diff
@@ -49,14 +52,31 @@
...
+    public CrumbIssuer getCrumbIssuer() {
+        return Jenkins.get().getCrumbIssuer();
+    }
+
+    @Restricted(NoExternalUse.class)
+    public static CrumbIssuer createDefaultCrumbIssuer() {
+        if (DISABLE_CSRF_PROTECTION) {
+            return null;
+        }
+        return new DefaultCrumbIssuer(SystemProperties.getBoolean(Jenkins.class.getName() + ".crumbIssuerProxyCompatibility",false));
```
Comment (jglick) [inline core/src/main/java/hudson/security/csrf/GlobalCrumbIssuerConfiguration.java]:
```suggestion
        return new DefaultCrumbIssuer(SystemProperties.getBoolean(Jenkins.class.getName() + ".crumbIssuerProxyCompatibility", false));
```

**Review by jglick — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Review by fcojfernandez — APPROVED** [review]
seems pretty good

**Review by alecharp — APPROVED** [review]
Great improvement indeed!

**Comment by timja** [comment]
6 reviews, last call for any negative feedback, we'll merge this tomorrow otherwise,
thanks Daniel for this!

**Review by oleg-nenashev — APPROVED** [review]
I will go ahead with the merge

**Comment by larrycai** [comment]
it happens here as well using helm chart to install with default value

**Comment by caseybea** [comment]
This issue (the crumbissuer error) happened here when upgrading Jenkins from version jenkins-2.235.5-1.1 to jenkins 2.249.1-1.1 on centos7.   This came from the jenkins-stable repo on Centos7.   It causes jenkins to not be able to start and I had to downgrade to the previous version to keep jekins running.

**Comment by mmerrill3** [comment]
I use the JCasC plugin, running jenkins in k8s as a container, and I need to disable the CSRF crumb issuer with javaOpts: "-Dhudson.security.csrf.GlobalCrumbIssuerConfiguration.DISABLE_CSRF_PROTECTION=true".  This is b/c the crumb check fails on plugins that I need to be installed, and prevents upgrading of plugins.  The end result is a jenkins install in a state that I did not declare with the configuration as code yaml.  For instance, older versions of plugins from previous installs.  I will watch JENKINS-63385 for updates for when I can disable this option.

**Pull request merged**